---
title: "101-set-up"
output: html_document
date: "2023-04-28"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
```

Putting data in 
$HOME/data/epv-safe-harbor/30-818141807    
sym linking to data/raw/


## Set up for analyses

Setting up FARM for analyses. Recent update has set a few things back. Checking for working modules
 module load STAR    
ERROR: Unable to locate a modulefile for 'STAR'     

(base) maccamp@farm:~$ STAR       
Usage: STAR  [options]... --genomeDir /path/to/genome/index/   --readFilesIn R1.fq R2.fq     

Great! I have installed this at some point.    
`rsync -a -P rsync://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.29_GRCh38.p14/ .`    

`mkdir STAR`     

`(base) maccamp@farm:~/genomes/hg38$ srun -p bigmemh -t 24:00:00 --mem=300GB --nodes=1 STAR --runThreadN 5 --runMode genomeGenerate --genomeDir ./STAR --genomeFastaFiles GCA_000001405.29_GRCh38.p14_genomic.fna --sjdbGTFfile GCA_000001405.29_GRCh38.p14_genomic.gtf`      

Salmon?

`(base) maccamp@farm:~$ module load salmon`    
Loading salmon/1.9.0    

`(base) maccamp@farm:~$ module load snakemake`     
ERROR: Unable to locate a modulefile for 'snakemake'     

conda install -n base -c conda-forge mamba
conda activate base
mamba create -c conda-forge -c bioconda -n snakemake snakemake


## diamond check

Build db in reference/diamond

```{sh, eval=FALSE}
diamond makedb --in seqs.faa -d seqs
```

  
header<-c("Query","Subj","%id","Length","Mismatch","Gap-open","Qstart","Qend","Sstart","Send","e-value","Bitscore") 
Working command diamond blastx -d seqs -q ~/Downloads/AA SRR17216319.fastq.gz \
  -o AVVS1-2-diamond 

working  command 
srun -p high -t 10:00:00 --ntasks=8 --cpus-per-task=1 --mem-per-cpu=50G parallel -j 8 < diamond-commands.txt

```{r}
meta<-read_tsv("meta/rna-seq-meta.csv")
new<-meta %>% filter(Donor %in% c("Donor.5","Donor.6","Donor.7"))
```

```{r}
command<-new %>% mutate(Command=paste0("diamond blastx -d  reference/diamond/seqs -q ", "data/raw/",
                                       Seqfile1, " -o outputs/101/",Sample,"-diamond"))
write_tsv(command %>% select(Command), file="101.1-new-diamond.sh", col_names = FALSE)
```

Considering a batch command

```{r}
write_tsv(new %>% select(Sample), col_names = FALSE, file="meta/samples.txt")
```

set up 101.2-diamond-batch.slurm to try

have 9 samples, so

sbatch -p med -J diamond.$USER --array=1-9 101.2-diamond-batch.slurm;

