---
title: "101-set-up"
output: html_document
date: "2023-04-28"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
```

Putting data in 
$HOME/data/epv-safe-harbor/30-818141807    
sym linking to data/raw/


## Set up for analyses

Setting up FARM for analyses. Recent update has set a few things back. Checking for working modules
 module load STAR    
ERROR: Unable to locate a modulefile for 'STAR'     

(base) maccamp@farm:~$ STAR       
Usage: STAR  [options]... --genomeDir /path/to/genome/index/   --readFilesIn R1.fq R2.fq     

Great! I have installed this at some point.    
`rsync -a -P rsync://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.29_GRCh38.p14/ .`    

`mkdir STAR`     

`(base) maccamp@farm:~/genomes/hg38$ srun -p bigmemh -t 24:00:00 --mem=300GB --nodes=1 STAR --runThreadN 5 --runMode genomeGenerate --genomeDir ./STAR --genomeFastaFiles GCA_000001405.29_GRCh38.p14_genomic.fna --sjdbGTFfile GCA_000001405.29_GRCh38.p14_genomic.gtf`      

Salmon?

`(base) maccamp@farm:~$ module load salmon`    
Loading salmon/1.9.0    

`(base) maccamp@farm:~$ module load snakemake`     
ERROR: Unable to locate a modulefile for 'snakemake'     

conda install -n base -c conda-forge mamba
conda activate base
mamba create -c conda-forge -c bioconda -n snakemake snakemake


## diamond check

Build db in reference/diamond

```{sh, eval=FALSE}
diamond makedb --in seqs.faa -d seqs
```

  
header<-c("Query","Subj","%id","Length","Mismatch","Gap-open","Qstart","Qend","Sstart","Send","e-value","Bitscore") 
Working command diamond blastx -d seqs -q ~/Downloads/AA SRR17216319.fastq.gz \
  -o AVVS1-2-diamond 

working  command 
srun -p high -t 10:00:00 --ntasks=8 --cpus-per-task=1 --mem-per-cpu=50G parallel -j 8 < diamond-commands.txt

```{r}
meta<-read_csv("meta/rna-seq-meta.csv")
new<-meta %>% filter(Donor %in% c("Donor.5","Donor.6","Donor.7"))
```

```{r}
command<-new %>% mutate(Command=paste0("diamond blastx -d  reference/diamond/seqs -q ", "data/raw/",
                                       Seqfile1, " -o outputs/101/",Sample,"-diamond"))
write_tsv(command %>% select(Command), file="101.1-new-diamond.sh", col_names = FALSE)
```

Considering a batch command

```{r}
samples<-new %>% select(Sample)
samples$Sample<-gsub("-","_",samples$Sample)
write_tsv(samples, col_names = FALSE, file="meta/samples.txt")
```

set up 101.2-diamond-batch.slurm to try

have 9 samples, so

sbatch -p bigmemm -J diamond.$USER --array=1-9 101.2-diamond-batch.slurm;

Is running!!

 cat Control-2-diamond | grep Mammalian  | awk '{ if ($4 > 44 && $3 > 99) { print } }' | wc -l
22

for f in *diamond; do echo $f; cat $f | grep Mammalian  | awk '{ if ($4 > 44 && $3 > 99) { print } }' | wc -l; done;

(base) maccamp@farm:~/epv-safe-harbor/outputs/101$ for f in *diamond; do echo $f; cat $f | grep Mammalian  | awk '{ if ($4 > 40 && $3 > 99) { print } }' | wc -l; done;
Control-1-diamond
0
Control-2-diamond
29
Control-3-diamond
0
NUP-1-diamond
345
NUP-2-diamond
229
NUP-3-diamond
1647
PVT-1-diamond
229
PVT-2-diamond
100
PVT-3-diamond
35

(base) maccamp@farm:~/epv-safe-harbor/outputs/101$ for f in *diamond; do echo $f; cat $f | grep Mammalian  | awk '{ if ($4 > 44 && $3 > 99) { print } }' | wc -l; done;
Control-1-diamond
901
Control-2-diamond
22
Control-3-diamond
92
NUP-1-diamond
311
NUP-2-diamond
188
NUP-3-diamond
1510
PVT-1-diamond
196
PVT-2-diamond
93
PVT-3-diamond
30

 
Let's compare NUP vs PVT for potential chimeras... Redoing on 10-24-2023 with new sequence data.
samples are 6
sbatch -p bigmemm -J diamond.$USER --array=1-6 101.2-diamond-batch.slurm;

Controls not working

srun -p high --nodes=1 --mem=32G -t 12:00:00 diamond blastx -d reference/diamond/seqs -q data/raw/Control_1_R1_0001.fastq.gz -o outputs/101/Control_1-diamond.tsv

srun -p high --nodes=1 --mem=32G -t 12:00:00 diamond blastx -d reference/diamond/seqs -q data/raw/Control_2_R1_0001.fastq.gz -o outputs/101/Control_2-diamond.tsv

cat NUP1-diamond.tsv | grep Mammalian  | awk '{ if ($4 > 20 && $4<44 && $3 > 99) { print } }'



for f in *diamond.tsv; do echo $f; cat $f | grep Mammalian  | awk '{ if ($4 > 44 && $3 > 99) { print } }' | wc -l; done;
Control_1-diamond.tsv
0
Control_2-diamond.tsv
0
NUP1-diamond.tsv
5443
NUP2-diamond.tsv
6787
PVT1-diamond.tsv
1306
PVT2-diamond.tsv
731

And a different alignment size
for f in *diamond.tsv; do echo $f; cat $f | grep Mammalian  | awk '{ if ($4 > 20 && $4<44 && $3 > 99) { print } }' | wc -l; done;
(base) maccamp@farm:~/epv-safe-harbor/outputs/101$ for f in *diamond.tsv; do echo $f; cat $f | grep Mammalian  | awk '{ if ($4 > 20 && $4<44 && $3 > 99) { print } }' | wc -l; done;

Control_1-diamond.tsv
0
Control_2-diamond.tsv
0
NUP1-diamond.tsv
1572
NUP2-diamond.tsv
2067
PVT1-diamond.tsv
410
PVT2-diamond.tsv
262

